src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -649,14 +657,25 @@ private static string GetUriWithRemovedQueryParameter(string uri, string name)
             }
         }
 
-        return newQueryStringBuilder.UriWithQueryString;
+        return newQueryStringBuilder.GetUriWithQueryString();
     }
 
     private static string GetUriWithAppendedQueryParameters(
         string uriWithoutQueryString,
         IReadOnlyDictionary<string, object?> parameters)
     {
-        var builder = new QueryStringBuilder(uriWithoutQueryString);
+        var uriWithoutQueryStringAndHash = uriWithoutQueryString;
+        var hash = "";
```

Comment:
Instead of `hash` being a `string`, could it be a `ReadOnlySpan<char>`? If it was populated using `uriWithoutQueryString.AsSpan()[hashStartIndex..]` then it wouldn't cause any heap allocations.

Same for the other `hash` variable inside `GetUriWithAppendedQueryParameter`, and the call to `uriWithoutQueryStringAndHash.Substring`